node/network/statement-distribution/src/lib.rs

```diff
@@ -911,38 +1098,249 @@ async fn report_peer(
+) -> Option<(&'a mut ActiveHeadData, SignedFullStatement)> {
+
+	let metadata = message.get_metadata();
+
+	let active_head = match active_heads.get_mut(&metadata.relay_parent) {
 		Some(h) => h,
 		None => {
 			tracing::debug!(
 				target: LOG_TARGET,
-				?peer,
-				?relay_parent,
-				"Unknown (or outdated) relay parent"
+				requested_relay_parent = %metadata.relay_parent,
+				"our view out-of-sync with active heads; head not found",
 			);
 			report_peer(ctx, peer, COST_UNEXPECTED_STATEMENT).await;
-			return None;
+			return None
+		}
+	};
+
+	match message {
+		protocol_v1::StatementDistributionMessage::Statement(_, statement) =>
+			return Some((active_head, statement)),
+
+		protocol_v1::StatementDistributionMessage::LargeStatement(metadata) => {
+
+			let candidate_hash = metadata.candidate_hash;
+
+			match active_head.waiting_large_statements.entry(candidate_hash) {
+				Entry::Occupied(mut occupied) => {
+
+					let mut new_status = None;
+
+					match occupied.get_mut() {
+						LargeStatementStatus::Fetching(info) => {
+							info.available_peers.insert(peer, metadata);
+							info.peers_to_try.push(peer);
+							// Answer any pending request for more peers:
+							if let Some(sender) = std::mem::take(&mut info.peer_sender) {
+								let to_send = std::mem::take(&mut info.peers_to_try);
+								if let Err(peers) = sender.send(to_send) {
+									// Requester no longer interested for now, might want them
+									// later:
+									info.peers_to_try = peers;
+								}
+							}
+						}
+						LargeStatementStatus::Fetched(committed) => {
+
+							let validator_id = active_head.validators.get(metadata.signed_by.0 as usize);
+
+							if let Some(validator_id) = validator_id {
+								let signing_context = SigningContext {
+									session_index: active_head.session_index,
+									parent_hash: metadata.relay_parent,
+								};
+
+								let statement = SignedFullStatement::new(
+									Statement::Seconded(committed.clone()),
+									metadata.signed_by,
+									metadata.signature.clone(),
+									&signing_context,
+									validator_id,
+								);
+
+								if let Some(statement) = statement {
+									return Some((active_head, statement));
+								} else {
+									// Cached data did not match signature, some backer wants to
+									// get slashed. At this point, we have to clear the cache and
+									// fetch the data again from the current peer. (We don't know
+									// which one is correct and it does not really matter, we leave
+									// that up for disputes. We should just try our best to make
+									// sure all valid and invalid data is available so it can be
+									// disputed and the offenders can be slashed.)
+									tracing::info!(
+										target: LOG_TARGET,
+										validator_index = ?metadata.signed_by,
+										"Building statement failed - invalid signature!"
+									);
+									new_status = launch_request(metadata, peer, req_sender.clone(), ctx).await;
```

Comment:
I still don't follow why we request again. Isn't this branch only triggered once a `launch_request` has already concluded?

Reply:
Yes, but if the signature does not match now, we should give the peer an opportunity to provide data that matches the signature. We store data only by `candidate_hash` so different `Seconded` messages will map to the same `LargeStatementStatus::Fetched`, this is important as collators send collations to multiple validators and we will get multiple `Seconded` messages. Those should be equivalent if the `candidate_hash` matches, but they might not, if the backer wants to get slashed - this code path handles that case.

Reply:
in any case, that's racy as the original peer might itself already have dropped the data. Will fix that.

Reply:
I was over-complicating things, we just need to verify the hash.